### DISCUSSION TITLE
Get an element of a variable as float

### DISCUSSION THREAD

**Discussion by mrrezaie** [casadi/casadi#3547] [Q&A] [answered]
Hi, I have a variable `A = opti.variable(n,m)` and want to pass one of its elements into a function (from another module) which requires a float as input. I found `__float__` method but `A[0,0].__float__()` doesn't work:
`RuntimeError: .../casadi/core/mx_node.cpp:386: 'to_double' not defined for class N6casadi16GetNonzerosSliceE`. 

I wonder if this conversion is possible. 

Any help is greatly appreciated.
-Mohammadreza

**Comment by jgillis** [comment, marked as answer]
You can add external purely numerical code in an ootimization by using a Callback, see the users guide ay docs.casadi.org

**Comment by mrrezaie** [comment]
Thanks for your response. Indeed I need to pass every element of `A` into that external function. This is the form that I'm looking for (merely an example):
```python
# A is an Opti variable
row,col = A.shape
for i in range(row):
    for j in range(col):
        B = a_non_casadi_function( A[i,j] ) # requires a float and returns a float
        opti.subject_to( B * A[i,j] == 0)
```
Any idea what would be the best way of doing this? I saw some examples of `Callback` and `Function` for getting numerical value(s) of specific element(s), but I don't know how to adapt them here (sorry I'm a beginner). In `scipy.optimize.minimize` I could define an equality constraint function and work with every element of that variable as a float object. I also know that in CasADi I need to work in a symbolic way. Any recommendation? Thanks again.

**Comment by mrrezaie** [comment]
Hi again, I could write something with `Callback`:

```python
# Define a custom function that calls a non-casadi function
class ExternalFunction(casadi.Callback):
    def __init__(self, name, opts={}):
        casadi.Callback.__init__(self)
        self.construct(name, opts)

    # Number of inputs and outputs
    def get_n_in(self): return 5
    def get_n_out(self): return 2

    # Initialize the object
    def init(self): print('Initializing the object')

    # Evaluate the function numerically
    def eval(self, arg):

        A = arg[0]
        B = arg[1]
        C = arg[2]
        D = arg[3]
        E = arg[4]

        # some numerical processing with functions from other modules
        # the output, F and G, are list of numbers

        return F, G
````

and 
```Python
f = ExternalFunction('f')

for i in range(N):
    out1, out2= f(var1[i], var2[i], var3[i], var4[i], var5[i])
    opti.subject_to(out1 == out2)
```

Do you think this structure is correct?

However, I get this error:

`RuntimeError: Error in Opti::solve [OptiNode] at .../casadi/core/optistack.cpp:157: .../casadi/core/function_internal.cpp:146: Error calling IpoptInterface::init for 'solver': Error in Function::call for 'nlp' [MXFunction] at .../casadi/core/function.cpp:339: Error in Function::operator() for 'f' [CallbackInternal] at .../casadi/core/function.cpp:1491: .../casadi/core/function_internal.cpp:2010: 'eval_sx' not defined for CallbackInternal`

Any help is greatly appreciated, thank you.

**Reply by jgillis** [reply]
Almost there. You must use MX symbols for your optimization (i.e. don't set `expand` `True`)

**Reply by mrrezaie** [reply]
I do appreciate your help, It is working very well in my simplified example. In my main workflow, I get an error and will post it if I couldn't find a solution. Thanks again for your help.

**Comment by jgillis** [comment]
I created a FAQ entry for your question: https://github.com/casadi/casadi/wiki/FAQ:-How-to-use-a-purely-numerical-function-inside-CasADi%3F-a.k.a-how-to-convert-a-symbol-to-a-number,-a.k.a-custom-functions,-a.k.a-Callback

**Comment by mrrezaie** [comment]
Hi, sorry for continuing this conversation. In my more elaborated model, callback rises this error:

`RuntimeError: Error in Opti::solve [OptiNode] at .../casadi/core/optistack.cpp:157:
.../casadi/core/function_internal.cpp:146: Error calling IpoptInterface::init for 'solver':
Error in Function::factory for 'nlp' [MXFunction] at .../casadi/core/function.cpp:1758:
Failed to create nlp_grad:[x, p, lam:f, lam:g]->[f, g, grad:gamma:x, grad:gamma:p] with {"gamma": [f, g]}:
.../casadi/core/factory.hpp:584: Gradient generation failed:
Error in MX::gradient at .../casadi/core/generic_matrix.hpp:1468:
Error in MX::jtimes at .../casadi/core/generic_matrix.hpp:1456:
Error in MX::reverse at .../casadi/core/mx.cpp:1788:
Error in MXFunction::ad_reverse at .../casadi/core/mx_function.cpp:1101:
Error in MX::ad_reverse for node of type N6casadi4CallE at .../casadi/core/mx.cpp:2310:
Error in Call::ad_reverse for 'f' [CallbackInternal] at .../casadi/core/casadi_call.cpp:149:
.../casadi/core/function_internal.cpp:2984: Assertion "has_derivative()" failed:
Derivatives cannot be calculated for f`

Then I added `enable_fd` argument when calling initializing the function as follows, but it is drastically slowing down the solver, i.e., every iteration takes ~1 min.
```python
f = ExternalFunction('f', {"enable_fd":True})
```

I saw the https://github.com/casadi/casadi/blob/main/docs/examples/python/callback.py examples, but don't know how to add additional functions in the Callback to fix this. Any idea/recommendation how to deal with this?

I do appreciate your help. Thank you.

**Reply by mrrezaie** [reply]
It would be great if you update this [FAQ ](https://github.com/casadi/casadi/wiki/FAQ:-How-to-use-a-purely-numerical-function-inside-CasADi%3F-a.k.a-how-to-convert-a-symbol-to-a-number,-a.k.a-custom-functions,-a.k.a-Callback) to address what to do if `has_derivative()` failed. Thank you.

**Comment by mrrezaie** [comment]
Hi again, at this time I decided not to use Callback because of the complexity. Now, I prepared a set of x and y values the external function gave me and defined several interpolant functions:
```python
AFLC = casadi.interpolant('AFLC', 'linear', [x1], y1)
```
Then, I put slices of my variables into these functions and finally I could define my constraints with them. But it doesn't work well and the solver is maximizing the objective rather than minimizing. Any idea what is wrong with this method?

(This is still related to this topic I think)

Any help is greatly appreciated.
